library/common/network/configurator.cc

```diff
@@ -229,9 +236,15 @@ void Configurator::refreshDns(envoy_netconf_t configuration_key, bool drain_conn
 
   if (auto dns_cache = dnsCache()) {
     ENVOY_LOG_EVENT(debug, "netconf_refresh_dns", std::to_string(configuration_key));
-    pending_drain_ = drain_connections && enable_drain_post_dns_refresh_;
-    // TODO(goaway): capture the list of currently cached hosts here in a set, to be checked
-    // for draining when DNS resolutions occur.
+
+    if (drain_connections && enable_drain_post_dns_refresh_) {
+      dns_cache->iterateHostMap(
+          [&](absl::string_view host,
+              const Extensions::Common::DynamicForwardProxy::DnsHostInfoSharedPtr&) {
+            hosts_to_drain_.emplace(host);
```

Comment:
should we explicitly empty `hosts_to_drain_` before we start adding hosts here? I think that we cannot end up in a state where we have dups in here but what if we had a host in the past that does not exist anymore (not sure whether this can happen)? 


Imagine we had DNS refresh 1) that included host x and y and DNS refresh 2) that included y host only. We want `hosts_to_drain_` to contain only `y` after the second DNS refresh is initialized

Reply:
I'm not actually sure if hosts can ever cease to exist in Envoy's cache implementation, since at the end of their TTL, I think they're simply refreshed. That said, previous re-resolutions should still exist even if this is triggered again before prior resolutions complete - in other words, I think the race where a host might persist exists irrespective of whether we call `clear` here.

In the case you describe:
The resolutions from 1) should still exist and be outstanding even if 2) is triggered, and thus we should still observe resolutions for x and y.

Reply:
Is it possible that 1) DNS refresh is cancelled and we never receives completion callback associated with it?

Reply:
My understanding is the callback we're listening for should fire regardless of the result state of the resolution. @mattklein123 added it because other callbacks didn't necessarily.